napari/_vispy/layers/labels.py

```diff
@@ -164,11 +176,51 @@ def hash2d_set(key, value, keys, values, empty_val=0):
             raise OverflowError('too many labels')
         key += 1
         pos = idx_to_2D(key, keys.shape)
-    keys[pos] = key
+    keys[pos] = initial_key
     values[pos] = value
 
 
-def build_textures_from_dict(color_dict, empty_val=0, shape=(1000, 1000)):
+def _get_shape_from_dict(color_dict):
+    size = len(color_dict) * 4
+    # I think that hash table size should be at least 4 times
+    # bigger than the number of labels to avoid collisions
+    for i, prime in enumerate(PRIME_NUM_TABLE[:-1]):
+        if prime * prime > size:
+            return prime, prime
+        if prime * PRIME_NUM_TABLE[i + 1] > size:
+            return PRIME_NUM_TABLE[i + 1], prime
+
+    if size > PRIME_NUM_TABLE[-1] * PRIME_NUM_TABLE[-1]:
+        raise OverflowError('too many labels')
+    return PRIME_NUM_TABLE[-1], PRIME_NUM_TABLE[-1]
+
+
+def get_shape_from_dict(color_dict):
+    global MAX_TEXTURE_SIZE
+    if MAX_TEXTURE_SIZE is None:
+        MAX_TEXTURE_SIZE = get_max_texture_sizes()[0]
+
+    shape = _get_shape_from_dict(color_dict)
+
+    if MAX_TEXTURE_SIZE is not None and (
+        shape[0] > MAX_TEXTURE_SIZE or shape[1] > MAX_TEXTURE_SIZE
+    ):
+        raise OverflowError(
+            f'Too many labels. GPU does not support textures of this size.'
+            f' Requested size is {shape[0]}x{shape[1]}, but maximum supported'
+            f' size is {MAX_TEXTURE_SIZE}x{MAX_TEXTURE_SIZE}'
```

Comment:
I think maybe this should be a MemoryError? OverflowError is a subclass of ArithmeticError, which is not what happened here, but rather a failure to allocate a large enough texture?

Reply:
I just use the same error as exists in code for this context. We could even implement one exception (`ToBigForShader` for example). What did you think?

Reply:
I generally dislike custom exceptions because then you need to import something to handle them. Personally I would go with a MemoryError here.

Reply:
Then, I prefer to create such an exception inheriting from an existing, less specific one. But I changed it to MemoryError.